src/main/kotlin/net/maizegenetics/phgv2/cli/Hvcf2Gvcf.kt

```diff
@@ -0,0 +1,571 @@
+
+        // Verify the tiledbURI - verifyURI will throw an exception if the URI is not valid
+        val validDB = verifyURI(dbPath,"hvcf_dataset",condaEnvPrefix)
+
+        // read and store ref file data for later use when creating the vcf sequence dictionary
+        val refSeq = CreateMafVcf().buildRefGenomeSeq(referenceFile)
+        buildGvcfFromHvcf(dbPath, refSeq, outputDir, hvcfDir, condaEnvPrefix)
+    }
+
+    fun buildGvcfFromHvcf(dbPath: String, refSeq: Map<String, NucSeq>, outputDir: String, hvcfDir: String, condaEnvPrefix: String) {
+        // get list of hvcf files
+        // walk the gvcf directory process files with g.vcf.gz extension
+        File(hvcfDir).walk().filter { !it.isHidden && !it.isDirectory }
+            .filter { it.name.endsWith("h.vcf.gz")  || it.name.endsWith("h.vcf")  }.toList()
+            .forEach { hvcfFile ->
+                myLogger.info("buildGvcfFromHvcf: Processing hvcf file: ${hvcfFile.name}")
+                val records = processSingleHVCF(refSeq, outputDir, hvcfFile, dbPath,condaEnvPrefix)
+
+                val sample = hvcfFile.toString().substringAfterLast("/").substringBefore(".")
+                val gvcfFile = "$outputDir/${sample}.g.vcf"
+                exportVariantContext(sample,records,gvcfFile, refSeq,setOf())
+            }
+
+    }
+
+    fun processSingleHVCF(refSeq:Map<String,NucSeq>, outputDir:String, hvcfFile: File, dbPath: String, condaEnvPrefix: String): List<VariantContext> {
+
+        val reader = VCFFileReader(hvcfFile,false)
+        // We also need to print to the new gvcf all the headers from 1 of the accessed gvcf files
+        val header = reader.fileHeader
+        val altHeaders = parseALTHeader(header = header)
+        val sampleNames = altHeaders.values.map { it.sampleName() }.toSet()
+        val contigNames = header.contigLines.map { it.id } // contig names needed for sorting
+
+        // This checks for existing gvcf files, and if they don't exist, exports them
+        val exportSuccess = exportGvcfFiles(sampleNames, outputDir, dbPath, condaEnvPrefix)
+
+        // Using the hvcfdFileReader, walk the hvcf file and for each entry create
+        // a ReferenceRange object and add that to a list of Reference Range objects
+        // for the sample to which it applies.  You will get the sample by indexing the
+        // altHeaders map with the key of the current record's ID.
+        val sampleToRefRanges = mutableMapOf<String, MutableList<ReferenceRange>>()
+
+        // Get the sampleName from the hvcf file.  This is the name
+        // without the path, and without the extension of h.vcf, hvcf, h.vcf.gz or hvcf.gz
+        val pathSample = hvcfFile.toString().substringAfterLast("/").substringBefore(".")
+
+        var rangesSkipped = 0
+        // process the hvcf records into the sampleToRefRanges map
+        reader.forEach { context ->
```

Comment:
Consider using a map{} so you can reduce the mutability of sampleToRefRanges.

Reply:
recoded with map{}